phoenix-core/src/main/java/org/apache/phoenix/coprocessor/CompactionScanner.java

```diff
@@ -0,0 +1,776 @@
+                    // Inside the max lookback window
+                    return true;
+                }
+                if (rowContext.maxTimestamp >= ttlWindowStart) {
+                    // Outside the max lookback window but inside the TTL window
+                    if (keepDeletedCells == KeepDeletedCells.FALSE &&
+                            dm.getTimestamp() < maxLookbackWindowStart ) {
+                        return false;
+                    }
+                }
+                if (keepDeletedCells == KeepDeletedCells.TTL &&
+                        dm.getTimestamp() >= ttlWindowStart) {
+                    return true;
+                }
+                if (maxLookbackInMillis != 0 &&
+                        dm.getTimestamp() > maxLookbackWindowStart && rowContext.version == 0) {
+                    // The delete marker is inside the max lookback window and this is the first
+                    // cell version outside the max lookback window. This cell should be
+                    // visible through the scn connection with the timestamp >= the delete
+                    // marker's timestamp
+                    return true;
+                }
+                return false;
+            }
+            i++;
+        }
+        return true;
+    }
+
+    /**
+     * Form the next compaction row version by picking the first cell from each column if the cell
+     * is not a delete marker (Type.Delete or Type.DeleteColumn) until the first delete family or
+     * delete family version column marker is visited
+     * @param columns
+     * @param rowContext
+     * @return
+     */
+    private CompactionRowVersion formNextCompactionRowVersion(List<List<Cell>> columns,
+            RowContext rowContext, boolean regionLevel) {
+        CompactionRowVersion rowVersion = null;
+        getNextRowVersionTimestamp(columns, rowContext);
+        for (List<Cell> column : columns) {
+            Cell cell = column.get(0);
+            if (cell.getType() == Cell.Type.DeleteFamily) {
+                if (cell.getTimestamp() >= rowContext.maxTimestamp) {
+                    rowContext.familyDeleteMarker = cell;
+                    column.remove(0);
+                }
+                continue;
+            }
+            else if (cell.getType() == Cell.Type.DeleteFamilyVersion) {
+                if (cell.getTimestamp() >= rowContext.maxTimestamp) {
+                    rowContext.familyVersionDeleteMarker = cell;
+                    column.remove(0);
+                }
+                continue;
+            }
+            if (rowContext.maxTimestamp != 0 &&
+                    column.get(0).getTimestamp() < rowContext.minTimestamp) {
+                continue;
+            }
+            column.remove(0);
+            if (cell.getType() == Cell.Type.DeleteColumn ||
+                    cell.getType() == Cell.Type.Delete) {
+                rowContext.addColumnDeleteMarker(cell);
```

Comment:
We add a column delete marker when we hit a row delete marker? (Cell.Type.Delete?)

Reply:
Cell.Type.Delete is a column delete marker to delete only one cell version. It should have been named Cell.Type.DeleteColumnVersion.